### ISSUE TITLE
Unhandled Exception: PlatformException(storekit_duplicate_product_object, There is a pending transaction for the same product identifier

### ISSUE THREAD

**Issue by rossman-x** [flutter/flutter#79473]
## Steps to Reproduce

1. Include in app purchase official plugin.
2. From sign up page on your iphone, create a payment button to redirect users to the in app purchase.
3. Create a subscription to follow purchase progress. 
4. First payment will work will, but if you cancel the purchase and try to pay again, it will show this error: `  Unhandled Exception: PlatformException(storekit_duplicate_product_object, There is a pending transaction for the same product identifier. Please either wait for it to be finished or finish it manually using 'completePurchase' to avoid edge cases., {applicationUsername: xxxx, requestData: null, quantity: 1, productIdentifier: XXXX, simulatesAskToBuyInSandbox: false}, null) `

I complete all purchases before and after running the purchase, using `completePurchase(details)`

<details> 
<summary>flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel stable, 2.0.3, on macOS 11.2 20D64 darwin-x64, locale en-GB)
    • Flutter version 2.0.3 at /Users/ramiosman/flutter
    • Framework revision 4d7946a68d (13 days ago), 2021-03-18 17:24:33 -0700
    • Engine revision 3459eb2436
    • Dart version 2.12.2

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/ramiosman/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.54.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (2 available)
    • Rami’s iPhone (mobile) • 00008030-00014D023A99802E • ios            • iOS 14.4.1
    • Chrome (web)           • chrome                    • web-javascript • Google Chrome 89.0.4389.90
    ! Error: iPhone is not connected. Xcode will continue when iPhone is connected. (code -13)
    ! Error: iPhone is not connected. Xcode will continue when iPhone is connected. (code -13)

! Doctor found issues in 1 category.
```
	
</details>

in_app_purchase version : ^0.5.1+1

Code : 

Listener : 
```dart
      _subscription = _iap.purchaseUpdatedStream
          .listen((List<PurchaseDetails> purchases) async {
        print(purchases.toString());
        if (purchases.isEmpty) return;
        await PaymentFutures.completePurchases(_iap);
        PurchaseDetails lastPurchase = purchases.last;
        if (lastPurchase.status == PurchaseStatus.error) {
          if (FirebaseAuth.instance.currentUser != null) {
            try {
              await counterRef
                  .child(FirebaseAuth.instance.currentUser.uid)
                  .set(null);
              await FirebaseAuth.instance.currentUser.delete();
            } catch (error) {
              print(error.toString());
            }
          }
          moveForward(true, isLoad: false);
          GlobalUseWidgets.alertDialog(context, "Payment error",
              "An error has occurred while trying to make the payment, please try again or contact support.");
          return;
        }

        if (lastPurchase.status == PurchaseStatus.purchased) {
          Navigator.of(context).push(MaterialPageRoute(builder: (bc) {
            return HomeScreen();
          }));
        }
        print(purchases.toString());
      });
      PaymentFutures.completePurchases(_iap).then((value) async {
        setState(() {
          isLoading = false;
          isInAppPurchaseAvailable = isAvailable;
        });
      });
```

PaymentFutures.completePurchases : 

```dart
  static Future<void> completePurchases(iap) async{
    final QueryPurchaseDetailsResponse purchaseDetailsResponse =
    await iap.queryPastPurchases();

    for (PurchaseDetails details in purchaseDetailsResponse.pastPurchases) {
      final pending = Platform.isIOS
          ? details.pendingCompletePurchase
          : !details.billingClientPurchase.isAcknowledged;
      if (pending) {
          await iap.completePurchase(details);
      }
    }
  }
```

runPurchase

```dart
  static Future<void> runPurchase(
      InAppPurchaseConnection iap, String uid) async {

    final ProductDetailsResponse productDetailsResponse =
        await iap.queryProductDetails({R.subscriptionIAPId});
    if (productDetailsResponse.productDetails.isEmpty)
      throw Future.error("Subscription not found");

    final ProductDetails details = productDetailsResponse.productDetails.first;

    final PurchaseParam purchaseParam = PurchaseParam(
        productDetails: details,
        applicationUserName: uid,
        sandboxTesting: false);

    if(Platform.isIOS) {
      var transactions = await SKPaymentQueueWrapper().transactions();
      transactions.forEach((skPaymentTransactionWrapper) {
        SKPaymentQueueWrapper().finishTransaction(skPaymentTransactionWrapper);
      });
    }

    await iap.buyNonConsumable(purchaseParam: purchaseParam);

  }
```

`await PaymentFutures.completePurchases(_iap);` is called before purchase is made.

**Comment by iapicca** [comment]
possibly realated to https://github.com/flutter/flutter/issues/60763

**Comment by TahaTesser** [comment]
@rossman-x 
Closing as duplicate of  https://github.com/flutter/flutter/issues/60763
Please follow up on that issue, If you disagree, please write in the comments and I will reopen it.
Thank you

**Comment by rossman-x** [comment]
@TahaTesser this is not a duplicate of #60763, @mvanbeusekom asked me to create a new issue (check last comments).

**Comment by rossman-x** [comment]
Using this simple flutter screen, 
```
class TestScreen extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MyHomePage();
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  static Set<String> _kIds = {'prod'};
  final InAppPurchaseConnection _storeConnection =
      InAppPurchaseConnection.instance;
  StreamSubscription<List<PurchaseDetails>> _subscription;
  bool _isAvailable;
  List<ProductDetails> _products = [];

  @override
  void initState() {
    final Stream<List<PurchaseDetails>> purchaseUpdates =
        _storeConnection.purchaseUpdatedStream;
    _subscription = purchaseUpdates.listen(
      (purchases) {
        _handlePurchaseUpdates(purchases);
      },
      onDone: () => _subscription.cancel(),
      onError: (error) => print('Purchase update error: $error'),
    );

    _initStore();
    super.initState();
  }

  Future<void> _initStore() async {
    bool isAvailable = await _storeConnection.isAvailable();

    if (!isAvailable) {
      setState(() {
        _isAvailable = isAvailable;
      });
      return;
    }

    final ProductDetailsResponse response =
        await _storeConnection.queryProductDetails(_kIds);

    if (response.error != null) {
      print('Error downloading product information: ${response.error}');

      setState(() {
        _isAvailable = isAvailable;
      });
    }

    setState(() {
      _isAvailable = isAvailable;
      _products = response.productDetails;
    });
  }

  @override
  Widget build(BuildContext context) {
    if (_products.isEmpty) {
      return const Center(
        child: Text('No products available.'),
      );
    }

    return Scaffold(
      appBar: AppBar(
        title: Text('IAP Demo'),
      ),
      body: ListView(
        children: [
          ..._buildProductList(),
        ],
      ),
    );
  }

  @override
  void dispose() {
    _subscription.cancel();
    super.dispose();
  }

  List<ListTile> _buildProductList() {
    return _products.map(
      (ProductDetails productDetails) {
        return ListTile(
            title: Text(
              productDetails.title,
            ),
            trailing: TextButton(
              child: Text(productDetails.price),
              style: TextButton.styleFrom(
                backgroundColor: Colors.green[800],
                primary: Colors.white,
              ),
              onPressed: () {
                PurchaseParam purchaseParam = PurchaseParam(
                  productDetails: productDetails,
                  applicationUserName: null,
                );

                _storeConnection.buyNonConsumable(purchaseParam: purchaseParam);
              },
            ));
      },
    ).toList();
  }

  Future<void> _handlePurchaseUpdates(
      List<PurchaseDetails> purchaseDetails) async {
    for (final purchase in purchaseDetails) {
      print('Received purchase update for:');
      print('\t- Product ID: ${purchase.productID}');
      print('\t- Purchase ID: ${purchase.purchaseID}');
      print('\t- Status: ${purchase.status}');

      if (purchase.pendingCompletePurchase) {
        await InAppPurchaseConnection.instance.completePurchase(purchase);
      }
    }
  }
}

```
After the first purchase I restart the application and tried to subscribe again, this is when the error shows up in my console log.

**Comment by rossman-x** [comment]
Actually I find out that running the code above will not cause any error, and the real issue is when this function is called : 
` final QueryPurchaseDetailsResponse purchaseDetailsResponse = await _storeConnection.queryPastPurchases(); `
try to put this line before you run the purchase and you will see this error: 

```

flutter: [Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails']
flutter: []
[VERBOSE-2:ui_dart_state.cc(186)] Unhandled Exception: PlatformException(storekit_duplicate_product_object, There is a pending transaction for the same product identifier. Please either wait for it to be finished or finish it manually using `completePurchase` to avoid edge cases., {applicationUsername: null, requestData: null, quantity: 1, productIdentifier: wsfsub, simulatesAskToBuyInSandbox: false}, null)
#0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:581:7)
#1      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:158:18)
<asynchronous suspension>
#2      SKPaymentQueueWrapper.addPayment (package:in_app_purchase/src/store_kit_wrappers/sk_payment_queue_wrapper.dart:88:5)
<asynchronous suspension>
#3      AppStoreConnection.buyNonConsumable (package:in_app_purchase/src/in_app_purchase/app_store_connection.dart:68:5)
<asynchronous suspension>
```

When I removed this line from code, it worked very well, except that now I cannot check (with a server call) if user is already subscribed or not ! which is absolutely not a solution ! this needs to be fixed !

**Comment by TahaTesser** [comment]
Hi @rossman-x 
Thanks for the details

**Comment by AmitabhWork** [comment]
> Actually I find out that running the code above will not cause any error, and the real issue is when this function is called :
> `final QueryPurchaseDetailsResponse purchaseDetailsResponse = await _storeConnection.queryPastPurchases();`
> try to put this line before you run the purchase and you will see this error:
> 
> ```
> 
> flutter: [Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails', Instance of 'PurchaseDetails']
> flutter: []
> [VERBOSE-2:ui_dart_state.cc(186)] Unhandled Exception: PlatformException(storekit_duplicate_product_object, There is a pending transaction for the same product identifier. Please either wait for it to be finished or finish it manually using `completePurchase` to avoid edge cases., {applicationUsername: null, requestData: null, quantity: 1, productIdentifier: wsfsub, simulatesAskToBuyInSandbox: false}, null)
> #0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:581:7)
> #1      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:158:18)
> <asynchronous suspension>
> #2      SKPaymentQueueWrapper.addPayment (package:in_app_purchase/src/store_kit_wrappers/sk_payment_queue_wrapper.dart:88:5)
> <asynchronous suspension>
> #3      AppStoreConnection.buyNonConsumable (package:in_app_purchase/src/in_app_purchase/app_store_connection.dart:68:5)
> <asynchronous suspension>
> ```
> 
> When I removed this line from code, it worked very well, except that now I cannot check (with a server call) if user is already subscribed or not ! which is absolutely not a solution ! this needs to be fixed !

Is there any workaround for this?

**Comment by AmitabhWork** [comment]
Please move this issue to label P3 @cyanglaz , without resolving this issue we cannot ship the app to app store, every time I call 
final QueryPurchaseDetailsResponse purchaseDetailsResponse = await _storeConnection.queryPastPurchases(); 
and after that try to call 
 _connection.buyNonConsumable(purchaseParam: purchaseParam);
 gets this error:
Unhandled Exception: PlatformException(storekit_duplicate_product_object, There is a pending transaction for the same product identifier. Please either wait for it to be finished or finish it manually using `completePurchase` to avoid edge cases., {applicationUsername: null, requestData: null, quantity: 1, productIdentifier: wsfsub, simulatesAskToBuyInSandbox: false}, null)

**Comment by rossman-x** [comment]
How can this bug be a P4 ? without fixing it all the package is useless on iPhone, please move it to p3 ! 
Also I have noticed something new about it, when I restart the app (Not hot restart), it tooks a while to `completePurchase` of all 104 purchases I have, why these purchases are still marked as pending even though I completed them in the run just before, why should I confirm all purchases each time I run my app ? the problem is that we cannot run a new purchase if we don't have all old purchases completed, the same error of pending purchase will show up.

**Comment by AmitabhWork** [comment]
> How can this bug be a P4 ? without fixing it all the package is useless on iPhone, please move it to p3 !
> Also I have noticed something new about it, when I restart the app (Not hot restart), it tooks a while to `completePurchase` of all 104 purchases I have, why these purchases are still marked as pending even though I completed them in the run just before, why should I confirm all purchases each time I run my app ? the problem is that we cannot run a new purchase if we don't have all old purchases completed, the same error of pending purchase will show up.

Yes Exact @rossman-x  same problem I am also facing, also you stated that if you don't use 
final QueryPurchaseDetailsResponse purchaseDetailsResponse = await _storeConnection.queryPastPurchases();
if works, but for me, it works until one fail purchase, after that, it will not work then I have to call _storeConnection.queryPastPurchases(); and wait for all the previous purchase to be verified, and then I have to remove _storeConnection.queryPastPurchases();  code restart the app and I can buyNonConsumable product again.

**Comment by cyanglaz** [comment]
Thank you for your patience in advance :) We are actively working on this issue. 
Basically we are going to introduce a new API `restorePastPurchase` and deprecate `queryPastPurchase`. The `restorePastPurchase` won't return anything, instead, the stream gets all the purchases. This will make the API less confusing and fix the issue during this change. 

FYI: This PR https://github.com/flutter/plugins/pull/3781 is part of the federation work that we are working on and somewhere down the line, the fix will be introduced.

**Comment by DigitalSolomon** [comment]
Thanks for the update, @cyanglaz . Perhaps someone can update the documentation on pub.dev to mention that `queryPastPurchases()` is being deprecated to prevent others from going too far down this rabbit hole.

**Comment by AmitabhWork** [comment]
> Thank you for your patience in advance :) We are actively working on this issue.
> Basically we are going to introduce a new API `restorePastPurchase` and deprecate `queryPastPurchase`. The `restorePastPurchase` won't return anything, instead, the stream gets all the purchases. This will make the API less confusing and fix the issue during this change.
> 
> FYI: This PR [flutter/plugins#3781](https://github.com/flutter/plugins/pull/3781) is part of the federation work that we are working on and somewhere down the line, the fix will be introduced.

As this issue is closed, may I know in which version these changes are added?

**Comment by mvanbeusekom** [comment]
@AmitabhWork these changes are part of [version 0.6.0](https://pub.dev/packages/in_app_purchase/versions/0.6.0), which has been released on May 13th.

**Comment by AmitabhWork** [comment]
> @AmitabhWork these changes are part of [version 0.6.0](https://pub.dev/packages/in_app_purchase/versions/0.6.0), which has been released on May 13th.

Hi, I upgraded to 0.6.0 but I am still getting the same issue.
I am still using the old sandbox testing ID.


Unhandled Exception: PlatformException(storekit_duplicate_product_object, There is a pending transaction for the same product identifier. Please either wait for it to be finished or finish it manually using `completePurchase` to avoid edge cases., {applicationUsername: amitabhs@sathguru.com, requestData: null, quantity: 1, productIdentifier: STANDARD_M_14022, simulatesAskToBuyInSandbox: false}, null)
#0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:581:7)
#1      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:158:18)
<asynchronous suspension>
#2      SKPaymentQueueWrapper.addPayment (package:in_app_purchase_ios/src/store_kit_wrappers/sk_payment_queue_wrapper.dart:91:5)
<asynchronous suspension>
#3      InAppPurchaseIosPlatform.buyNonConsumable (package:in_app_purchase_ios/src/in_app_purchase_ios_platform.dart:63:5)
<asynchronous suspension>

**Comment by mvanbeusekom** [comment]
@AmitabhWork this is possible. There probably are still old transaction in the payment queue which you should complete first.

In general these transactions are delivered (since version 0.6.0) through the `InAppPurchase.purchaseStream` with a `status` of `PurchaseStatus.restored`. This is where you should verify them, deliver the content (if valid) and always (valid or not) complete them by calling `InAppPurchase.completePurchase` method and supplying the purchase.

If you don't receive these purchases on the `InAppPurchase.purchaseStream`, you can still complete them using the `SKPaymentQueueWrapper` class. Below is an example:

```dart
import 'package:in_app_purchase_ios/in_app_purchase_ios.dart';

var transactions = await SKPaymentQueueWrapper().transactions();
transactions.forEach((skPaymentTransactionWrapper) {
  SKPaymentQueueWrapper().finishTransaction(skPaymentTransactionWrapper);
});
```

**Comment by AmitabhWork** [comment]
> ```dart
> SKPaymentQueueWrapper()
> ```
version 0.6.0

import 'package:in_app_purchase_ios/in_app_purchase_ios.dart';
var transactions = await SKPaymentQueueWrapper().transactions();

The method 'SKPaymentQueueWrapper' isn't defined for the type '_InAppPurchaseIOSState'.

SKPaymentQueueWrapper class is undefined, what am I doing wrong?


is there a way to catch this exception Unhandled Exception: PlatformException(storekit_duplicate_product_object

**Comment by AmitabhWork** [comment]
> @AmitabhWork this is possible. There probably are still old transaction in the payment queue which you should complete first.
> 
> In general these transactions are delivered (since version 0.6.0) through the `InAppPurchase.purchaseStream` with a `status` of `PurchaseStatus.restored`. This is where you should verify them, deliver the content (if valid) and always (valid or not) complete them by calling `InAppPurchase.completePurchase` method and supplying the purchase.
> 
> If you don't receive these purchases on the `InAppPurchase.purchaseStream`, you can still complete them using the `SKPaymentQueueWrapper` class. Below is an example:
> 
> ```dart
> import 'package:in_app_purchase_ios/store_kit_wrappers.dart';
> 
> var transactions = await SKPaymentQueueWrapper().transactions();
> transactions.forEach((skPaymentTransactionWrapper) {
>   SKPaymentQueueWrapper().finishTransaction(skPaymentTransactionWrapper);
> });
> ```
> 
> **UPDATED**:
> 
> I noticed the example contained a small mistake. Make sure you import `package:in_app_purchase_ios/store_kit_wrappers.dart` instead of `package:in_app_purchase_ios/in_app_purchae_ios.dart`.

Issue not resolved, I am getting this issue in production also, Is there any way to catch this exception?
even after restoring purchase and after using this  SKPaymentQueueWrapper().finishTransaction(skPaymentTransactionWrapper); the issue doesn't resolve it get the same error Unhandled Exception: PlatformException(storekit_duplicate_product_object..... , I am just not able to go through purchase.

**Comment by mvanbeusekom** [comment]
@AmitabhWork to help debug this problem I would like to check a few things if possible.

Do you make sure all purchases that arrive in the `InAppPurchase.purchaseStream` are completed? Ideally the `listen` callback would look like this:

```dart
void _listenToPurchaseUpdated(List<PurchaseDetails> purchaseDetailsList) {
  purchaseDetailsList.forEach((PurchaseDetails purchaseDetails) async {
    if (purchaseDetails.status == PurchaseStatus.pending) {
      // Indicate that the purchase is still pending...
    } else {
      if (purchaseDetails.status == PurchaseStatus.error) {
        // Handle the error.
      } else {
        // Validate the receipt and if valid deliver content.
      }
    }

    // Make sure the purchase always is completed if the `pendingCompletePurchase`
    // is set to `true`.
    if (purchaseDetails.pendingCompletePurchase) {
      await InAppPurchase.instance
          .completePurchase(purchaseDetails);
    }
  });
}
``` 

Would it be possible for you to share the code of your callback method?

Second I want to verify if you use the `restorePurchases` method in your code? Especially on iOS you should only use this when the user indicates to start restoring purchases (e.g. by putting the functionality behind a button). A common mistake is that people call the `restorePurchases` method when initialising their App. If you do this (which is against Apples guidelines) make sure you do it after you have subscribed to the `InAppPurchase.purchaseStream` stream.

On iOS each restored purchase will result in a new transaction. Each of these transactions need to be completed. If you restore purchases before you have subscribed to the `InAppPurchase.purchaseStream` there is a change those transactions will arrive late (e.g. next time you restart your App, which again restores purchases and again create new transactions) and in the meantime you are not allowed to purchase the same product again (resulting in the `PlatformException(storekit_duplicate_product_object....` exception).


 >  Is there any way to catch this exception?
 
 Yes, you can. This exception would occur when you are calling the `buyConsumable`  or `buyNonConsumable` methods, so you should be able to add a `try..catch` around that method call and catch the exception. Something like this:
 
 ```dart
 try {
  await InAppPurchase.instance.buyNonConsumable(purchaseParam);
 } on PlatformException catch (e) {
   if (e.code == 'storekit_duplicate_product_object') {
     // Handle the exception here...
   }
 }
 ```

**Comment by BaranMichal25** [comment]
@mvanbeusekom 
I realise this situation is somewhat a developer's mistake, but I think that the error should be handled and delivered via `purchaseStream` or it should be stated in documentation of method `buyNonConsumable` that it can throw an exception 🤔

**Comment by Hwan-seok** [comment]
I faced the same issue and solved it as follows
```dart
on InAppPurchase.instance.purchaseStream.listen ...

case PurchaseStatus.error:
    await InAppPurchase.instance.completePurchase(purchaseDetails);
```

**Comment by heilmela** [comment]
I'm facing the same issue when using `restorePurchases`. I suspect a timing issue, while i receive `PurchaseStatus.restored` in the `updateListener` and call `completePurchase` the queue is blocked by a duplicate when calling `buyNonConsumable`. Interestingly waiting some seconds and then retrying works despite the log does not show any pending transactions (to be completed) in the `purchaseListener()`. 

Note that i use `SKPaymentQueueWrapper` (https://github.com/flutter/flutter/issues/79473#issuecomment-842245471) to successfully finish a number of transactions before calling `buyNonConsumable`, which does not resolve the issue. 

This is my `purchaseUpdateListener`

```
Future<void> _handlePurchaseUpdate(
    List<PurchaseDetails> purchaseDetailsList,
  ) async {
    final restores = purchaseDetailsList
        .where((PurchaseDetails details) =>
            details.status == PurchaseStatus.restored)
        .toList();

    purchaseDetailsList.forEach((PurchaseDetails purchaseDetails) async {
      if (purchaseDetails.status == PurchaseStatus.pending) {
        ensureProcessingPurchase();
      } else {
        if (purchaseDetails.status == PurchaseStatus.error) {
          store.dispatch(PurchaseCancel());
        } else if (purchaseDetails.status == PurchaseStatus.restored) {
          //noop
          //only the last restored purhcase is verified below
        } else {
          ensureProcessingPurchase();
          _verifyAndFinishTransaction(purchaseDetails);
        }
      }

      if (purchaseDetails.pendingCompletePurchase) {
        await InAppPurchase.instance.completePurchase(purchaseDetails);
        print('Completed Purchase');
      }
    });

    if (restores.isNotEmpty) {
      _restoreLastPurchase(restores);
    } 
  }
```
